File: server-go/hub/router.go
package hub

import (
	"encoding/binary"
	"encoding/json"
	"log"
	"net"
	"net/http"
	"os"
	"path/filepath"

	"github.com/gorilla/websocket"
)

var upgrader = websocket.Upgrader{
	CheckOrigin: func(r *http.Request) bool { return true },
}

func corsMiddleware(next http.HandlerFunc) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Access-Control-Allow-Origin", "*")
		w.Header().Set("Access-Control-Allow-Methods", "GET, POST, OPTIONS")
		w.Header().Set("Access-Control-Allow-Headers", "Content-Type")
		if r.Method == http.MethodOptions {
			w.WriteHeader(http.StatusOK)
			return
		}
		next(w, r)
	}
}

func InitRouter() {
	// 1. Rust Agent 接入通道
	http.HandleFunc("/agent/register", handleAgentRegister)

	// 2. 面向第三方系统的开放 API 接口
	http.HandleFunc("/api/v1/devices", corsMiddleware(handleListDevices))
	http.HandleFunc("/api/v1/devices/thumbnail", corsMiddleware(handleThumbnail))
	http.HandleFunc("/api/v1/stream", handleStreamSubscribe)
	http.HandleFunc("/api/v1/control", corsMiddleware(handleExternalControl))
	http.HandleFunc("/agents", corsMiddleware(handleAgentDownload))
}

// 接收 Rust Agent 的画面数据并高效流式分发给所有第三方订阅者
func handleAgentRegister(w http.ResponseWriter, r *http.Request) {
	conn, err := upgrader.Upgrade(w, r, nil)
	if err != nil {
		log.Println("Upgrade error:", err)
		return
	}
	deviceID := r.URL.Query().Get("device_id")
	deviceName := r.URL.Query().Get("device_name")
	deviceOS := r.URL.Query().Get("os")
	deviceCPU := r.URL.Query().Get("cpu")
	deviceRAM := r.URL.Query().Get("ram")
	deviceMAC := r.URL.Query().Get("mac")
	deviceIP := r.RemoteAddr
	if deviceID == "" {
		conn.Close()
		return
	}

	GlobalHub.mu.Lock()
	GlobalHub.Agents[deviceID] = conn
	GlobalHub.DeviceInfos[deviceID] = DeviceInfo{
		ID:   deviceID,
		Name: deviceName,
		OS:   deviceOS,
		IP:   deviceIP,
		CPU:  deviceCPU,
		RAM:  deviceRAM,
		MAC:  deviceMAC,
	}
	GlobalHub.mu.Unlock()

	defer func() {
		GlobalHub.mu.Lock()
		delete(GlobalHub.Agents, deviceID)
		delete(GlobalHub.DeviceInfos, deviceID)
		GlobalHub.mu.Unlock()
		conn.Close()
	}()

	for {
		// 极速读取 Rust 发来的自定义二进制画面包，零解析，直接以字节数组形态向外广播
		messageType, payload, err := conn.ReadMessage()
		if err != nil {
			break
		}

		if messageType == websocket.BinaryMessage {
			if len(payload) > 14 && (payload[0] == 0x02 || payload[0] == 0x03 || payload[0] == 0x04) {
				GlobalHub.mu.Lock()
				// 保存完整帧（含 14 字节头），供新订阅者连接时直接推送
				frame := make([]byte, len(payload))
				copy(frame, payload)
				GlobalHub.LatestFrame[deviceID] = frame
				GlobalHub.mu.Unlock()
			}
			GlobalHub.mu.RLock()
			subs := GlobalHub.Subscribers[deviceID]
			for subConn := range subs {
				// 异步无阻塞流式转发原始二进制画面块
				err := subConn.WriteMessage(websocket.BinaryMessage, payload)
				if err != nil {
					log.Println("Write error to sub:", err)
				}
			}
			GlobalHub.mu.RUnlock()
		} else if messageType == websocket.TextMessage {
			displayLen := len(payload)
			if displayLen > 200 {
				displayLen = 200
			}
			log.Printf("[Hub] Agent %s sent text: %s", deviceID, string(payload[:displayLen]))
			GlobalHub.mu.RLock()
			subs := GlobalHub.Subscribers[deviceID]
			log.Printf("[Hub] Forwarding text to %d subscribers", len(subs))
			for subConn := range subs {
				subConn.WriteMessage(websocket.TextMessage, payload)
			}
			GlobalHub.mu.RUnlock()
		}
	}
}

// 接收外部控制 API 的 JSON 请求，直接秒级透传至 Rust 被控端执行物理模拟
func handleExternalControl(w http.ResponseWriter, r *http.Request) {
	// 确保任何情况下都有响应（避免浏览器 ERR_EMPTY_RESPONSE）
	w.Header().Set("Content-Type", "application/json")

	if r.Method != http.MethodPost {
		w.WriteHeader(http.StatusMethodNotAllowed)
		w.Write([]byte(`{"error":"method not allowed"}`))
		return
	}
	var cmd map[string]interface{}
	if err := json.NewDecoder(r.Body).Decode(&cmd); err != nil {
		w.WriteHeader(http.StatusBadRequest)
		w.Write([]byte(`{"error":"invalid json"}`))
		return
	}

	deviceID, ok := cmd["device_id"].(string)
	if !ok {
		w.WriteHeader(http.StatusBadRequest)
		w.Write([]byte(`{"error":"missing device_id"}`))
		return
	}

	// 先写 HTTP 响应（保证浏览器不报 ERR_EMPTY_RESPONSE），再异步透传到 agent
	w.WriteHeader(http.StatusOK)
	w.Write([]byte(`{"status":"success"}`))

	GlobalHub.mu.RLock()
	agentConn, exists := GlobalHub.Agents[deviceID]
	GlobalHub.mu.RUnlock()

	if exists {
		bytes, _ := json.Marshal(cmd)
		agentConn.WriteMessage(websocket.TextMessage, bytes)
	}
}

func handleListDevices(w http.ResponseWriter, r *http.Request) {
	GlobalHub.mu.RLock()
	list := make([]DeviceInfo, 0, len(GlobalHub.DeviceInfos))
	for _, info := range GlobalHub.DeviceInfos {
		if info.Name == "" {
			info.Name = info.ID
		}
		list = append(list, info)
	}
	GlobalHub.mu.RUnlock()

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(list)
}

func handleThumbnail(w http.ResponseWriter, r *http.Request) {
	deviceID := r.URL.Query().Get("device_id")
	if deviceID == "" {
		w.WriteHeader(http.StatusBadRequest)
		return
	}

	GlobalHub.mu.RLock()
	frame, exists := GlobalHub.LatestFrame[deviceID]
	GlobalHub.mu.RUnlock()

	if !exists || len(frame) < 15 {
		w.WriteHeader(http.StatusNotFound)
		w.Write([]byte("Thumbnail not available yet"))
		return
	}

	w.Header().Set("Content-Type", "image/jpeg")
	w.Header().Set("Cache-Control", "no-cache, no-store, must-revalidate")
	w.Header().Set("Access-Control-Allow-Origin", "*")
	// 跳过 14 字节头，输出纯 JPEG 数据
	w.Write(frame[14:])
}

func handleStreamSubscribe(w http.ResponseWriter, r *http.Request) {
	conn, err := upgrader.Upgrade(w, r, nil)
	if err != nil {
		log.Println("Upgrade error:", err)
		return
	}
	deviceID := r.URL.Query().Get("device_id")

	GlobalHub.mu.Lock()
	wasEmpty := len(GlobalHub.Subscribers[deviceID]) == 0
	if GlobalHub.Subscribers[deviceID] == nil {
		GlobalHub.Subscribers[deviceID] = make(map[*websocket.Conn]bool)
	}
	GlobalHub.Subscribers[deviceID][conn] = true
	GlobalHub.mu.Unlock()

	// 第一个订阅者连接时通知 Agent 恢复截图
	if wasEmpty {
		GlobalHub.mu.RLock()
		agentConn, exists := GlobalHub.Agents[deviceID]
		GlobalHub.mu.RUnlock()
		if exists {
			agentConn.WriteMessage(websocket.TextMessage, []byte(`{"action":"resume"}`))
		}
	}

	// 订阅时立即推送最新缓存帧，确保被控端画面静止时控制窗口也能显示画面
	GlobalHub.mu.RLock()
	latestFrame := GlobalHub.LatestFrame[deviceID]
	GlobalHub.mu.RUnlock()
	if latestFrame != nil {
		conn.WriteMessage(websocket.BinaryMessage, latestFrame)
	}

	defer func() {
		var becameEmpty bool
		GlobalHub.mu.Lock()
		if GlobalHub.Subscribers[deviceID] != nil {
			delete(GlobalHub.Subscribers[deviceID], conn)
			if len(GlobalHub.Subscribers[deviceID]) == 0 {
				delete(GlobalHub.Subscribers, deviceID)
				becameEmpty = true
			}
		}
		GlobalHub.mu.Unlock()
		conn.Close()

		// 最后一个订阅者离开时通知 Agent 暂停截图
		if becameEmpty {
			GlobalHub.mu.RLock()
			agentConn, exists := GlobalHub.Agents[deviceID]
			GlobalHub.mu.RUnlock()
			if exists {
				agentConn.WriteMessage(websocket.TextMessage, []byte(`{"action":"pause"}`))
			}
		}
	}()

	// 挂起连接持续等待监听退订事件和控制指令
	for {
		messageType, payload, err := conn.ReadMessage()
		if err != nil {
			break
		}

		// 将客户端发来的 WebSocket 文本消息（控制指令）透传给 Agent，降低控制延迟
		if messageType == websocket.TextMessage || messageType == websocket.BinaryMessage {
			if messageType == websocket.TextMessage {
				displayLen := len(payload)
				if displayLen > 200 {
					displayLen = 200
				}
				log.Printf("[Hub] Subscriber for %s sent text: %s", deviceID, string(payload[:displayLen]))
			}
			GlobalHub.mu.RLock()
			agentConn, exists := GlobalHub.Agents[deviceID]
			GlobalHub.mu.RUnlock()
			if exists {
				agentConn.WriteMessage(messageType, payload)
			} else {
				log.Printf("[Hub] Agent %s not found, cannot forward message", deviceID)
			}
		}
	}
}

var exeConfigGUID = []byte{0xB9, 0x96, 0x01, 0x58, 0x80, 0x54, 0x4A, 0x19, 0xB7, 0xF7, 0xE9, 0xBE, 0x44, 0x91, 0x4C, 0x18}

func getLocalIP() string {
	ifaces, err := net.Interfaces()
	if err != nil {
		return "127.0.0.1"
	}
	for _, iface := range ifaces {
		if iface.Flags&net.FlagUp == 0 || iface.Flags&net.FlagLoopback != 0 {
			continue
		}
		addrs, err := iface.Addrs()
		if err != nil {
			continue
		}
		for _, addr := range addrs {
			var ip net.IP
			switch v := addr.(type) {
			case *net.IPNet:
				ip = v.IP
			case *net.IPAddr:
				ip = v.IP
			}
			if ip != nil && ip.To4() != nil {
				return ip.String()
			}
		}
	}
	return "127.0.0.1"
}

func handleAgentDownload(w http.ResponseWriter, r *http.Request) {
	exePath := filepath.Join("..", "agent-rust", "target", "release", "exhibition-agent.exe")
	if _, err := os.Stat(exePath); os.IsNotExist(err) {
		exePath = filepath.Join("..", "agent-rust", "target", "debug", "exhibition-agent.exe")
		if _, err := os.Stat(exePath); os.IsNotExist(err) {
			w.WriteHeader(http.StatusNotFound)
			w.Write([]byte("Agent exe not found"))
			return
		}
	}

	host := r.URL.Query().Get("server")
	if host == "" {
		host = r.Host
	}
	port := "38921"
	if h, p, err := net.SplitHostPort(host); err == nil {
		host = h
		port = p
	}
	if host == "localhost" || host == "127.0.0.1" || host == "" {
		host = getLocalIP()
	}
	host = net.JoinHostPort(host, port)

	scheme := "http"
	if r.TLS != nil {
		scheme = "https"
	}
	configJSON, _ := json.Marshal(map[string]string{"server_url": scheme + "://" + host})

	exeData, err := os.ReadFile(exePath)
	if err != nil {
		w.WriteHeader(http.StatusInternalServerError)
		return
	}

	w.Header().Set("Content-Type", "application/octet-stream")
	w.Header().Set("Content-Disposition", "attachment; filename=\"exhibition-agent.exe\"")
	w.Write(exeData)
	w.Write(configJSON)
	binary.Write(w, binary.BigEndian, uint32(len(configJSON)))
	w.Write(exeConfigGUID)
}

File: Test001/server-node/hub/router.js

const http = require('http');
const fs = require('fs');
const path = require('path');
const WebSocket = require('ws');
const os = require('os');
const { GlobalHub } = require('./device-hub');
const { streamExeWithConfig, getAgentFilePath } = require('./exe-config');

function getLocalIP() {
  const interfaces = os.networkInterfaces();
  for (const name of Object.keys(interfaces)) {
    for (const iface of interfaces[name]) {
      if (iface.family === 'IPv4' && !iface.internal) {
        return iface.address;
      }
    }
  }
  return '127.0.0.1';
}

function initWSServer(server) {
  const wss = new WebSocket.Server({ noServer: true, perMessageDeflate: true });

  server.on('upgrade', (req, socket, head) => {
    wss.handleUpgrade(req, socket, head, ws => wss.emit('connection', ws, req));
  });

  // 被控端连接
  wss.on('connection', (ws, req) => {
    console.log(`WebSocket连接: ${req.url}`);
    if (req.url.includes('/agent/register')) {
      const params = new URLSearchParams(req.url.split('?')[1]);
      const deviceId = params.get('device_id');
      console.log(`Agent注册请求: device_id=${deviceId}`);
      if (!deviceId) return ws.close(1008, "invalid device id");

      const ip = req.socket.remoteAddress;
      GlobalHub.registerAgent(deviceId, ip, ws);

      ws.on('message', (data) => {
        // 心跳处理
        if (data.toString() === 'ping') {
          GlobalHub.updateHeartbeat(deviceId);
          return ws.send('pong');
        }

        // 二进制画面转发
        if (Buffer.isBuffer(data)) {
          GlobalHub.broadcastBinary(deviceId, data);
          return;
        }

        // 普通文本信令 / CMD终端指令回包 / 遥测数据
        try { 
          const msg = JSON.parse(data);
          GlobalHub.updateHeartbeat(deviceId);
          if (msg.action === 'register') {
            GlobalHub.updateDeviceInfo(deviceId, { name: msg.name, os_name: msg.os });
          }
          if (msg.action === 'telemetry') {
            GlobalHub.updateDeviceTelemetry(deviceId, msg);
          }
          // 转发文本消息到订阅者（如遥测、终端输出）
          GlobalHub.broadcastText(deviceId, data);
        } catch (e) {}
      });

      ws.on('close', () => GlobalHub.unregisterAgent(deviceId));
      ws.on('error', () => GlobalHub.unregisterAgent(deviceId));
    }

    // 画面订阅连接
    else if (req.url.includes('/api/v1/stream')) {
      const params = new URLSearchParams(req.url.split('?')[1]);
      const deviceId = params.get('device_id');
      console.log(`画面订阅请求: device_id=${deviceId}`);
      if (!deviceId) return ws.close(1008, "invalid device id");

      GlobalHub.addSubscriber(deviceId, ws);
      console.log(`流订阅成功: ${deviceId}, subscribers=${GlobalHub.subscribers.get(deviceId)?.size}`);
      
      // 浏览器→agent 消息转发（键鼠指令、质量调整等）
      ws.on('message', (data) => {
        if (Buffer.isBuffer(data) && data.length <= 10) {
          console.log(`input relay: ${deviceId}, len=${data.length}, type=0x${data[0]?.toString(16)}`);
        }
        GlobalHub.sendCmdPriority(deviceId, data);
      });
      
      ws.on('close', (code, reason) => {
        console.log(`流断开: ${deviceId}, code=${code}, reason=${reason?.toString()||'none'}`);
        GlobalHub.removeSubscriber(deviceId, ws);
      });
      ws.on('error', (err) => {
        console.log(`流错误: ${deviceId}, error=${err.message}`);
        GlobalHub.removeSubscriber(deviceId, ws);
      });
    }
  });

  return wss;
}

function handleApi(req, res) {
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'GET,POST,OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type');
  if (req.method === 'OPTIONS') return res.writeHead(200), res.end();

  // 设备列表
  if (req.url === '/api/v1/devices' && req.method === 'GET') {
    res.writeHead(200, { 'Content-Type': 'application/json' });
    return res.end(JSON.stringify(GlobalHub.listDevices()));
  }

  // Agent 下载端点
  if (req.url.startsWith('/agents') && req.method === 'GET') {
    const agentPath = getAgentFilePath();
    if (!agentPath) {
      res.writeHead(404, { 'Content-Type': 'text/plain' });
      return res.end('Agent executable not found');
    }

    const params = new URLSearchParams(req.url.split('?')[1]);
    let serverUrl = params.get('server');
    
    if (!serverUrl) {
      const host = req.headers.host || '';
      const port = host.includes(':') ? host.split(':')[1] : '80';
      const localIP = getLocalIP();
      serverUrl = `http://${localIP}:${port}`;
    }

    const config = {
      server_url: serverUrl,
      device_name: params.get('name') || undefined,
      width: params.get('width') ? parseInt(params.get('width')) : undefined,
      height: params.get('height') ? parseInt(params.get('height')) : undefined,
    };

    res.writeHead(200, {
      'Content-Type': 'application/octet-stream',
      'Content-Disposition': 'attachment; filename="exhibition-agent.exe"',
    });

    console.log(`Agent download requested, server_url: ${config.server_url}`);
    
    try {
      streamExeWithConfig({
        sourceFileName: agentPath,
        destinationStream: res,
        config: config,
      });
    } catch (e) {
      console.error('Agent download error:', e);
      res.writeHead(500, { 'Content-Type': 'text/plain' });
      res.end('Internal server error');
    }
    return;
  }

  // 通用控制指令下发（键鼠 + CMD终端指令）
  if (req.url === '/api/v1/control' && req.method === 'POST') {
    let body = '';
    req.on('data', c => body += c);
    req.on('end', () => {
      try {
        const cmd = JSON.parse(body);
        const ok = GlobalHub.sendCmdPriority(cmd.device_id, JSON.stringify(cmd));
        if (ok) {
          res.writeHead(200, { 'Content-Type': 'application/json' });
          res.end(JSON.stringify({ status: "success" }));
        } else {
          res.writeHead(404, { 'Content-Type': 'application/json' });
          res.end(JSON.stringify({ status: "device offline" }));
        }
      } catch {
        res.writeHead(500, { 'Content-Type': 'application/json' });
        res.end(JSON.stringify({ status: "server error" }));
      }
    });
    return;
  }

  // 静态文件服务：非 API 请求返回前端页面
  const frontendDir = path.join(__dirname, '..', 'public');
  let filePath = path.join(frontendDir, req.url === '/' ? 'index.html' : req.url);
  
  // 安全检查：防止路径遍历
  if (!filePath.startsWith(frontendDir)) {
    res.writeHead(403);
    res.end('Forbidden');
    return;
  }
  
  const ext = path.extname(filePath);
  const contentTypes = {
    '.html': 'text/html; charset=utf-8',
    '.js': 'application/javascript',
    '.css': 'text/css',
    '.json': 'application/json',
    '.png': 'image/png',
    '.jpg': 'image/jpeg',
    '.ico': 'image/x-icon',
  };
  
  fs.readFile(filePath, (err, data) => {
    if (err) {
      res.writeHead(404, { 'Content-Type': 'text/plain' });
      res.end('Not Found');
    } else {
      res.writeHead(200, { 'Content-Type': contentTypes[ext] || 'application/octet-stream' });
      res.end(data);
    }
  });
}

module.exports = { initWSServer, handleApi };
